From 1c12cf06e9132336b991c26bef87a9117e2e9ea9 Mon Sep 17 00:00:00 2001
From: Shubhrajyoti Datta <shubhrajyoti.datta@xilinx.com>
Date: Mon, 10 Feb 2020 10:52:12 +0530
Subject: [PATCH 1144/1851] tty: pl011: Add support for parity configuration

commit 84364d402403ecd64b9119d8ea51808298bc4650 from
https://github.com/Xilinx/linux-xlnx.git

Currently we are not updating the parity.
Add support for parity configuration.

Signed-off-by: Shubhrajyoti Datta <shubhrajyoti.datta@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/tty/serial/amba-pl011.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/tty/serial/amba-pl011.c b/drivers/tty/serial/amba-pl011.c
index 6642e2c54cf0..17f502574101 100644
--- a/drivers/tty/serial/amba-pl011.c
+++ b/drivers/tty/serial/amba-pl011.c
@@ -2074,7 +2074,6 @@ sbsa_uart_set_termios(struct uart_port *port, struct ktermios *termios,
 
 	tty_termios_encode_baud_rate(termios, uap->fixed_baud, uap->fixed_baud);
 	/* The SBSA UART only supports 8n1 without hardware flow control. */
-	termios->c_cflag &= ~(CSTOPB | PARENB | PARODD);
 	termios->c_cflag &= ~(CMSPAR | CRTSCTS);
 	switch (termios->c_cflag & CSIZE) {
 	case CS5:
@@ -2090,6 +2089,15 @@ sbsa_uart_set_termios(struct uart_port *port, struct ktermios *termios,
 		lcr_h = UART01x_LCRH_WLEN_8;
 		break;
 	}
+	if (termios->c_cflag & CSTOPB)
+		lcr_h |= UART01x_LCRH_STP2;
+	if (termios->c_cflag & PARENB) {
+		lcr_h |= UART01x_LCRH_PEN;
+		if (!(termios->c_cflag & PARODD))
+			lcr_h |= UART01x_LCRH_EPS;
+		if (termios->c_cflag & CMSPAR)
+			lcr_h |= UART011_LCRH_SPS;
+	}
 	if (uap->fifosize > 1)
 		lcr_h |= UART01x_LCRH_FEN;
 
-- 
2.31.1

